ctapipe/reco/impact.py

```diff
@@ -726,53 +878,81 @@ def predict(self, shower_seed, energy_seed):
         )
         horizon = nominal.transform_to(AltAz())
 
-        tilted = TiltedGroundFrame(
+        shower_result = ReconstructedGeometryContainer()
+        shower_result.prefix = self.__class__.__name__
+        # Transform everything back to a useful system
+        shower_result.alt, shower_result.az = horizon.alt, horizon.az
+        tilted = SkyCoord(
             x=fit_params[2] * u.m,
             y=fit_params[3] * u.m,
             z=0 * u.m,
-            pointing_direction=self.array_direction,
+            frame=TiltedGroundFrame(pointing_direction=self.array_direction),
         )
         ground = project_to_ground(tilted)
 
-        h_max = (
+        shower_result.core_x = ground.x
+        shower_result.core_y = ground.y
+
+        shower_result.core_tilted_x = tilted.x
+        shower_result.core_tilted_y = tilted.y
+
+        shower_result.core_tilted_uncert_x = errors[2] * u.m
+        shower_result.core_tilted_uncert_y = errors[3] * u.m
+
+        shower_result.is_valid = True
+
+        # Currently no errors not available to copy NaN
+        shower_result.alt_uncert = source_x * u.rad
+        shower_result.az_uncert = source_y * u.rad
+        # shower_result.core_uncert = np.nan
+
+        # Copy reconstructed Xmax
+        x_max = (
             fit_params[5]
-            * np.cos(zenith)
             * self.get_shower_max(
                 fit_params[0],
                 fit_params[1],
                 fit_params[2],
                 fit_params[3],
                 zenith.to(u.rad).value,
             )
+            * np.cos(zenith)
         )
-
-        shower_result = ReconstructedGeometryContainer(
-            alt=horizon.alt,
-            az=horizon.az,
-            core_x=ground.x,
-            core_y=ground.y,
-            is_valid=True,
-            alt_uncert=np.nan,
-            az_uncert=np.nan,
-            core_uncert=np.nan,
-            # Copy reconstructed hmax
-            h_max=h_max,
-            h_max_uncert=errors[5] * h_max,
-            goodness_of_fit=like,
-            prefix=self.__class__.__name__,
+        h_max = self.altitude_profile(x_max)
```

Comment:
I think this is likely incorrect: the old atmosphere profile functions only worked for vertical heights above sea level, and not distances thought the atmosphere. In other words, this only works correctly at zenith=0 degrees.   This should be replaced with the conversion functions provided by `ctapipe.atmosphere.AtmosphereDenistyProfile`.   However, currently we only have a `line_of_sight_integral(distance, zenith_angle)` function that converts *h_max* to *x_max*, so we will need to implement the inverse transform as well.  I'll open an issue and try to fix that. 

However, I notice that here x_max is defined in terms of cos(z), so it's a hight, not a distance thought the atmosphere.  We should also check elsewhere that we are consistent in it's definition: is it supposed to be the atmospheric depth along the shower's path (slant depth), or vertically to the shower-max (atmosphere depth)? I would think the former is more useful as it is the actual amount of atmosphere traversed by the shower's progenitor.  Or is the latter what the model expects?   In HillasReconstructor, I think we reconstruct the slant depth.

Reply:
I would think so as well.  But clearly here the hight only is being reconstructed.  Of course you can always compute the distance from the hight and core position, but I think it's better that we always assume h_max/x_max are slant depths.  

@gschwefer maybe check everywhere that x_max/h_max are used here that they are really using slant depth and not height, and fix places where they are not.  That could easily lead to poor behavior at larger zenith angles.  Maybe we could also add a unit test to reconstruct a dummy shower at small and large zenith angle (say 10º, 60°) and see if the results both come out correctly.

Reply:
Whenever x_max is used here it should be the slant depth at it is the only physically meaningful thing to use. I think this is just an error and instead we should use:
` h_max = self.altitude_profile(x_max*cos(zenith_angle))`
But also do we really want to store h_max in our output and not x_max?

Reply:
> But also do we really want to store h_max in our output and not x_max?

We can store both. h_max is already there because that is what the HillasReconstructor reconstructs geometrically;  x_max requires applying the atmosphere profile, so it was not included by default.  But we should always have a quite good estimate of the current atmosphere density, so no good reason not to just reconstruct x_max always.

Reply:
Ok, so I looked at this issue and as I see it everything in the branch is mathematically correct and consistent, though the naming is not as clear as it could be. Impact indeed intrinsically reconstructs slant depth. The value of h_max that is returned by the class is just the vertical height of the shower maximum, obtained through the altitude_profile function, which as an input correctly gets cos(zenith)*slant_depth.
The get_shower_max function correctly returns the guess of the slant depth of the shower maximum, by geometrically calculating the height above the telescope in the tilted frame, converting this to vertical height over ground (and sea level), using the thickness_profile function to get the column density from this and then divides by cos(zenith) to get slant depth.

I've renamed variables now so x_max is exclusively used for slant depth, the vertical depth is called column_density now. Any suggestions on what to name the h_max or is the name fine?

Reply:
`column_density` is actually just another name for slant depth - it implies a line-of-site integration, so that is somewhat confusing; I think from the previous discussion about this in another issue that the correct term is  "mass overburden". However,  I would avoid renaming or redefining anything in this pull request.  We can do that in a separate PR where we are careful to be consistent. 

Here, just keep with the definition of x_max and h_max as both representing the height/mass-overburden, and avoid using slant depth anywhere.

Reply:
Ok, I agree `column_density` is maybe not the ideal name. I also should have been clearer above: I only renamed internal variables used inside the `impact.py` file because the names were internally inconsistent, I didn't change anything globally or in the containers.

Reply:
In astrophysics it's the line-of-sight integral through a medium, for example in X-ray analysis through molecular clouds